### ISSUE TITLE
请教一下为什么要修改model.py中的模型结构代码？

### ISSUE THREAD

**Issue by oak-JHCheng** [DakeQQ/Automatic-Speech-Recognition-ASR-ONNX#12]
作者您好，非常感谢您的分享，整个项目的阅读和使用都很便捷！
在测试sensevoicesmall-onnx性能的过程中，我发现您对model.py中的部分网络结构有做修改，最直接的是SenseVoiceEncoderSmall类中新增了LFR_LENGTH: int = 100,FEATURE_SIZE: int = 560,USE_EMOTION: bool = True三个初始化参数进行位置编码的计算，并在forward中将该位置编码与输入相加；还有其它网络结构如EncoderLayerSANM和MultiHeadedAttentionSANM也都有一定程度的修改。
想问一下做这些修改的目的是什么呢，这种修改不会对原模型的效果有影响吗？期待您的解答

**Comment by DakeQQ** [comment]
针对SenseVoiceEncoderSmall的优化，我们为其类新增了三个初始化参数，用于预先计算位置编码。相较于原始代码中根据输入音频长度即时计算的方式，我们现在预先计算所有可能的位置编码，并在推理时通过低成本的Slice算子高效提取所需部分。

我们对SenseVoice的整体修改策略围绕两大核心：预计算和张量形状转置。这旨在确保模型在`forward()`推理阶段能以最少的算子完成操作，从而提升效率。

此外，我们还高度关注GPU算子的适配性，力求减少CPU与GPU之间不必要的频繁数据通信（例如ONNX Runtime编译时可能插入的MemcpyToHost/FromHost）。因此，提升GPU算子利用率并最大程度减少通信算子是另一重要优化目标。

最后，我们对修改后的代码进行了验证。通过对各语种输出内容的测试，目前尚未发现明显的异常。

**Comment by DakeQQ** [comment]
当使用`DYNAMIC_AXES = False`导出固定输入张量尺寸的模型时，确实会发生您担忧的情形“强制填充到固定长度”以及“过长音频被截断”。然而，您可以选择使用`DYNAMIC_AXES = True` ，并给予`INPUT_AUDIO_LENGTH`一个合理的大值（填入60*16000也是可以正常导出并推理的），如此推理流水线不会强制填充白噪音至固定长度，并且在小于`INPUT_AUDIO_LENGTH`内，摸型皆能动态接收不同長度的音频而不报错。填充白噪音在此处没有特别用意。填充白噪音主要是为了配合降噪模型，如果填充0的话，容易引起过度降噪末尾吃字的情况。

**Comment by oak-JHCheng** [comment]
> 当使用`DYNAMIC_AXES = False`导出固定输入张量尺寸的模型时，确实会发生您担忧的情形“强制填充到固定长度”以及“过长音频被截断”。然而，您可以选择使用`DYNAMIC_AXES = True` ，并给予`INPUT_AUDIO_LENGTH`一个合理的大值（填入60*16000也是可以正常导出并推理的），如此推理流水线不会强制填充白噪音至固定长度，并且在小于`INPUT_AUDIO_LENGTH`内，摸型皆能动态接收不同長度的音频而不报错。填充白噪音在此处没有特别用意。填充白噪音主要是为了配合降噪模型，如果填充0的话，容易引起过度降噪末尾吃字的情况。

是的您说的对，如果不考虑推理时间，将DYNAMIC_AXES = True，并给予INPUT_AUDIO_LENGTH一个合理的大值，可以满足大部分场景，但是我现在要把他用在实时转录场景中，对时延的要求比较高，所以要尽可能的缩短推理时长，这样就会出现我上面提到的问题1：“任何时长小于30s的音频，其特征长度都会被padding到与30s音频相同，当然时耗也与30s音频相同；”，这在实时转录场景下是不友好的。
但是我刚刚修改了一版您的代码，将onnx模型导出前，将需要“提前计算的参数”或“初始化阶段的计算操作”移动到了推理forward流程中，测试了一下是可行的：
1、可以输入不同长度的音频，并且音频长度越短，耗时越少
2、onnx模型的加速效果相比于修改前相差不多？
我这里提供给您一组测试数据：
**测试数据**
session_opts.inter_op_num_threads = 1    
session_opts.intra_op_num_threads = 1
音频长度  200次推理平均时长
5s             14.65ms
10s           19.73ms
20s           30.1ms
30s           37.41ms
此前INPUT_AUDIO_LENGTH=35s的onnx模型，推理30s音频的耗时是39.91ms
这样来看是否不需要提前指定INPUT_AUDIO_LENGTH来避免所带来的一些限制？
希望对您有帮助

**Comment by DakeQQ** [comment]
您的反馈测试结果非常有价值。不过，仔细阅读您的描述后，我发现其中可能存在一些误解。

您提到：
> "将DYNAMIC_AXES = True，并给予INPUT_AUDIO_LENGTH一个合理的大值...这样就会出现问题1：任何时长小于30s的音频，其特征长度都会被padding到与30s音频相同，当然时耗也与30s音频相同"

**我对您这段话的理解是**：
即使设置DYNAMIC_AXES = True，您仍然遇到了"任何时长小于30s的音频都会被padding到与30s音频相同长度"的情况。

**实际情况澄清**：
对于DYNAMIC_AXES = True的模型，音频预处理代码**不会**对小于INPUT_AUDIO_LENGTH的音频进行padding。模型接收的音频长度就是实际输入长度，推理耗时也相应调整。相反，如果输入音频长度超过INPUT_AUDIO_LENGTH，则会在ONNX Runtime内部报错并退出。

**关键代码分析**：

```python
# 对于DYNAMIC_AXES = True的模型，shape_value_in为字符串类型，必然进入if分支
# 这里只是做上限保护，防止ONNX Runtime内部报错
# 您可以随时修改具体数值，只需小于等于导出时的最大设定值
if isinstance(shape_value_in, str):
    INPUT_AUDIO_LENGTH = min(320000, audio_len)  # 当audio_len < 320000时，INPUT_AUDIO_LENGTH = audio_len
else:
    INPUT_AUDIO_LENGTH = shape_value_in

# ...后续处理...

# 由于INPUT_AUDIO_LENGTH = audio_len，代码不会进入以下任一分支，因此原音频不会进行任何填充
# 传递给模型的长度就是音频原始长度
# 如此短音频应不会增加额外的推理时间
if audio_len > INPUT_AUDIO_LENGTH:
    num_windows = int(np.ceil((audio_len - INPUT_AUDIO_LENGTH) / stride_step)) + 1
    total_length_needed = (num_windows - 1) * stride_step + INPUT_AUDIO_LENGTH
    pad_amount = total_length_needed - audio_len
    final_slice = audio[:, :, -pad_amount:].astype(np.float32)
    white_noise = (np.sqrt(np.mean(final_slice * final_slice)) * np.random.normal(loc=0.0, scale=1.0, size=(1, 1, pad_amount))).astype(audio.dtype)
    audio = np.concatenate((audio, white_noise), axis=-1)
elif audio_len < INPUT_AUDIO_LENGTH:
    audio_float = audio.astype(np.float32)
    white_noise = (np.sqrt(np.mean(audio_float * audio_float)) * np.random.normal(loc=0.0, scale=1.0, size=(1, 1, INPUT_AUDIO_LENGTH - audio_len))).astype(audio.dtype)
    audio = np.concatenate((audio, white_noise), axis=-1)
```

**关键点**：
- DYNAMIC_AXES = True时，`INPUT_AUDIO_LENGTH = min(320000, audio_len)`确保了短音频不会被填充
- 这种设计保证了DYNAMIC_AXES = True + 短音频段 不会因填充而增加不必要的推理时间

希望这个澄清能帮助您更好地理解代码逻辑。如果您在实际测试中仍然观察到不同的行为，请提供更详细的测试场景，我们可以进一步分析。

最后，非常欢迎您按需修改代码。我们将导出代码井然条列便于断点追踪，目的也是为了让码友方便修改。

**Comment by oak-JHCheng** [comment]
> 您的反馈测试结果非常有价值。不过，仔细阅读您的描述后，我发现其中可能存在一些误解。
> 
> 您提到：
> 
> > "将DYNAMIC_AXES = True，并给予INPUT_AUDIO_LENGTH一个合理的大值...这样就会出现问题1：任何时长小于30s的音频，其特征长度都会被padding到与30s音频相同，当然时耗也与30s音频相同"
> 
> **我对您这段话的理解是**： 即使设置DYNAMIC_AXES = True，您仍然遇到了"任何时长小于30s的音频都会被padding到与30s音频相同长度"的情况。
> 
> **实际情况澄清**： 对于DYNAMIC_AXES = True的模型，音频预处理代码**不会**对小于INPUT_AUDIO_LENGTH的音频进行padding。模型接收的音频长度就是实际输入长度，推理耗时也相应调整。相反，如果输入音频长度超过INPUT_AUDIO_LENGTH，则会在ONNX Runtime内部报错并退出。
> 
> **关键代码分析**：
> 
> # 对于DYNAMIC_AXES = True的模型，shape_value_in为字符串类型，必然进入if分支
> # 这里只是做上限保护，防止ONNX Runtime内部报错
> # 您可以随时修改具体数值，只需小于等于导出时的最大设定值
> if isinstance(shape_value_in, str):
>     INPUT_AUDIO_LENGTH = min(320000, audio_len)  # 当audio_len < 320000时，INPUT_AUDIO_LENGTH = audio_len
> else:
>     INPUT_AUDIO_LENGTH = shape_value_in
> 
> # ...后续处理...
> 
> # 由于INPUT_AUDIO_LENGTH = audio_len，代码不会进入以下任一分支，因此原音频不会进行任何填充
> # 传递给模型的长度就是音频原始长度
> # 如此短音频应不会增加额外的推理时间
> if audio_len > INPUT_AUDIO_LENGTH:
>     num_windows = int(np.ceil((audio_len - INPUT_AUDIO_LENGTH) / stride_step)) + 1
>     total_length_needed = (num_windows - 1) * stride_step + INPUT_AUDIO_LENGTH
>     pad_amount = total_length_needed - audio_len
>     final_slice = audio[:, :, -pad_amount:].astype(np.float32)
>     white_noise = (np.sqrt(np.mean(final_slice * final_slice)) * np.random.normal(loc=0.0, scale=1.0, size=(1, 1, pad_amount))).astype(audio.dtype)
>     audio = np.concatenate((audio, white_noise), axis=-1)
> elif audio_len < INPUT_AUDIO_LENGTH:
>     audio_float = audio.astype(np.float32)
>     white_noise = (np.sqrt(np.mean(audio_float * audio_float)) * np.random.normal(loc=0.0, scale=1.0, size=(1, 1, INPUT_AUDIO_LENGTH - audio_len))).astype(audio.dtype)
>     audio = np.concatenate((audio, white_noise), axis=-1)
> **关键点**：
> 
> * DYNAMIC_AXES = True时，`INPUT_AUDIO_LENGTH = min(320000, audio_len)`确保了短音频不会被填充
> * 这种设计保证了DYNAMIC_AXES = True + 短音频段 不会因填充而增加不必要的推理时间
> 
> 希望这个澄清能帮助您更好地理解代码逻辑。如果您在实际测试中仍然观察到不同的行为，请提供更详细的测试场景，我们可以进一步分析。
> 
> 最后，非常欢迎您按需修改代码。我们将导出代码井然条列便于断点追踪，目的也是为了让码友方便修改。

关于“任何时长小于30s的音频，其特征长度都会被padding到与30s音频相同，当然时耗也与30s音频相同”这一点，实际上是我的测试结果，以onnx-35s的模型对5s和30s的音频进行推理，时耗是相同的（均为40ms）。我的描述有一些错误，其中的"padding"并非指推理前的音频预处理，而是有两个参数在模型初始化时被固化为固定值（基于INPUT_AUDIO_LENGTH）：
STFT_SIGNAL_LENGTH = INPUT_AUDIO_LENGTH // HOP_LENGTH + 1
LFR_LENGTH = (STFT_SIGNAL_LENGTH + LFR_N - 1) // LFR_N
在class SENSE_VOICE和class SenseVoiceEncoderSmall(nn.Module)中作为初始参数
class SENSE_VOICE(torch.nn.Module):
    def __init__(self, sense_voice, stft_model, cif_hidden_size, nfft_stft, stft_signal_len, n_mels, sample_rate, pre_emphasis, lfr_m, lfr_n, **lfr_len**, cmvn_means, cmvn_vars, use_emo):
        super(SENSE_VOICE, self).__init__()
        ...
        indices = torch.arange(0, **self.T_lfr** * lfr_n, lfr_n, dtype=torch.int64).unsqueeze(1) + torch.arange(lfr_m, dtype=torch.int64)
        self.indices_mel = indices.clamp(max=**stft_signal_len** + self.lfr_m_factor - 1)
        ...
class SenseVoiceEncoderSmall(nn.Module):
    """
    Author: Speech Lab of DAMO Academy, Alibaba Group
    SCAMA: Streaming chunk-aware multihead attention for online end-to-end speech recognition
    https://arxiv.org/abs/2006.01713
    """

    def __init__(
            self,
            ...
            **LFR_LENGTH:** int = 100,
            FEATURE_SIZE: int = 560,
            USE_EMOTION: bool = True,
            **kwargs,
    ):
           ...
        positions = torch.arange(1, **LFR_LENGTH** + 5, device='cpu')[None, :].type(torch.float32) if USE_EMOTION else torch.arange(1, LFR_LENGTH + 4, device='cpu')[None, :].type(torch.float32)
        log_timescale_increment = torch.log(torch.tensor([10000], dtype=torch.float32, device='cpu')) / (FEATURE_SIZE / 2 - 1)
        inv_timescales = torch.exp(torch.arange(FEATURE_SIZE / 2, device='cpu').type(torch.float32) * (-log_timescale_increment))
        inv_timescales = torch.reshape(inv_timescales, [1, -1])
        scaled_time = torch.reshape(positions, [1, -1, 1]) * torch.reshape(inv_timescales, [1, 1, -1])
        self.position_encoding = torch.cat([torch.sin(scaled_time), torch.cos(scaled_time)], dim=2)
虽然没有确定是在哪一步短音频被扩展为了长音频，但我想和这两处有关？

**Comment by DakeQQ** [comment]
大致了解您的情况。对于动态尺寸且输入长度超过导出参数预设值的模型，确实会对超出的末尾进行填充，但这填充不是必要的，仅仅是代码逻辑缺陷。因为这段代码主要是给`DYNAMIC_AXES = False`使用的，对于动态尺寸的模型則相当于亡羊补牢。

我们当初设计的理念，是推测ASR大多伴随着VAD模型一块使用，而鲜少有人能连续不停输出活动语音达到30秒或以上而不被VAD截断。因此在设计模型的导出与推理流水线时加入了此考察，才有今天的代码结果。

**Comment by oak-JHCheng** [comment]
> 大致了解您的情况。对于动态尺寸且输入长度超过导出参数预设值的模型，确实会对超出的末尾进行填充，但这填充不是必要的，仅仅是代码逻辑缺陷。因为这段代码主要是给`DYNAMIC_AXES = False`使用的，对于动态尺寸的模型則相当于亡羊补牢。
> 
> 我们当初设计的理念，是推测ASR大多伴随着VAD模型一块使用，而鲜少有人能连续不停输出活动语音达到30秒或以上而不被VAD截断。因此在设计模型的导出与推理流水线时加入了此考察，才有今天的代码结果。
好的我了解了，因为刚刚接触onnx加速相关的知识，很多细节处理不理解，看来我还需要继续研读该项目代码，感谢您关于该话题的回答，我受益匪浅，随后我会关闭该issue，顺便请教一下，如果想要学习onnx模型加速相关的技术，对于初学者您有推荐的课程或书籍吗

**Comment by DakeQQ** [comment]
学习初期可以直接找LLM聊，因为这是很成熟的框架，LLM的回答基本没啥幻觉，比博客或书籍更能快速入门。但后期我觉得写的最清楚的还得是官方文档。
https://onnxruntime.ai/docs/
https://onnxruntime.ai/docs/api/python/api_summary.html
